db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/VerifyAggregateExpressionsVisitor.java

```diff
@@ -36,137 +36,142 @@
-				(node instanceof SubqueryNode) ||
-				(node instanceof ValueNode &&
-						groupByList != null 
-						&& groupByList.findGroupingColumn((ValueNode)node) != null));
-	}
-	
-	public boolean stopTraversal()
-	{
-		return false;
-	}
-
-	public boolean visitChildrenFirst(Visitable node)
-	{
-		return false;
-	}
+    public Visitable visit(Visitable node, QueryTreeNode parent) throws StandardException {
+        if (node instanceof ColumnReference) {
+            ColumnReference cr = (ColumnReference) node;
+
+            if (groupByList == null) {
+                throw StandardException.newException(SQLState.LANG_INVALID_COL_REF_NON_GROUPED_SELECT_LIST, cr.getSQLColumnName());
+            }
+
+            if (groupByList.findGroupingColumn(cr) == null) {
+                throw StandardException.newException(SQLState.LANG_INVALID_COL_REF_GROUPED_SELECT_LIST, cr.getSQLColumnName());
+            }
+        }
+
+        /*
+         ** Subqueries are only valid if they do not have
+         ** correlations and are expression subqueries.  RESOLVE:
+         ** this permits VARIANT expressions in the subquery --
+         ** should this be allowed?  may be confusing to
+         ** users to complain about:
+         **
+         ** select max(x), (select sum(y).toString() from y) from x
+         */
+        else if (node instanceof SubqueryNode) {
+            SubqueryNode subquery = (SubqueryNode) node;
+
+            if ((subquery.getSubqueryType() != SubqueryNode.EXPRESSION_SUBQUERY) ||
+                    subquery.hasCorrelatedCRs()) {
+                if (groupByList != null) {
+                    /*
+                     ** Check if the SSQ's correlated columns are referencing the grouping columns, if so, allow it.
+                     */
+                    List<ValueNode> correlationCRs = subquery.getCorrelationCRs();
+                    for (ValueNode correlationCR : correlationCRs) {
+                        if (correlationCR instanceof ColumnReference) {
+                            boolean included = false;
+                            for (OrderedColumn column : groupByList.getNodes()) {
+                                if (column.getValueNode().equals(correlationCR)) {
+                                    included = true;
+                                    break;
+                                }
+                            }
+                            if (!included) {
+                                throw StandardException.newException(SQLState.LANG_INVALID_GROUPED_SELECT_LIST);
+                            }
+                        } else {
+                            throw StandardException.newException(SQLState.LANG_INVALID_GROUPED_SELECT_LIST);
+                        }
+                    }
+                } else {
+                    throw StandardException.newException(SQLState.LANG_INVALID_NON_GROUPED_SELECT_LIST);
+                }
+            }
+
+            if (!subquery.hasCorrelatedCRs()) {
```

Comment:
I'm not sure why we cannot handle aggregate in the subquery, it seems perfectly fine for both scenarios below:
``` 
select a1, (select max(b2) from t2), count(*) from t1 group by a1;
select a1, (select max(b2) from t2 where a2=t1.a1), count(*) from t1 group by a1;
```
Can you lift this restriction?

Reply:
Hi Yi,

I was also wondering why subquery aggregations were, as commented, "temporarily" disabled. I thought I would keep it this way since addressing it is probably outside the scope of this PR. Anyway, I remove the restriction and added a test for it.

Reply:
Thanks!